File: typescript/src/generation.ts
/** Wire vocabulary shared by placement maps and native write barriers. */
import { MigrationRefused } from './errors.js'
export const DRAIN_TABLE = '__sde_fence_drains'
export const EPOCH_COLUMN = '__sde_write_epoch'
export const MAX_EPOCH = 9_007_199_254_740_991
/** The placement map contract that introduced `project_id` and per-group `write_epoch`. */
export const GENERATIONS_SINCE = 4
export function checkEpoch(epoch: number): number {
  if (!Number.isSafeInteger(epoch) || epoch < 1) {
    throw new MigrationRefused('write epoch must be a positive safe integer')
  }
  return epoch
}

import type { LogicalModel } from './model.js'
import type { PhysicalFinding } from './physical.js'
import { fingerprintOf, type PhysicalLayout, type PlacementMap } from './placement.js'
import type { Engine, Row } from './session.js'
import type { WriteFence } from './write-fence.js'

export interface Fencable {
  writeFence(table: string, options: { projectId: string }): WriteFence
  /**
   * Refuses missing columns and types; returns how tables differ from the declared physical
   * design when `keys` are given. `void` is accepted from an adapter written before findings.
   */
  validateSchema(
    layout: PhysicalLayout,
    options?: { readonly keys?: Readonly<Record<string, readonly string[]>> },
  ): Promise<readonly PhysicalFinding[] | void>
}

export function checkMapProject(placement: PlacementMap, projectId: string | undefined): string | undefined {
  if (placement.contract < 4) return undefined
  if (projectId === undefined || projectId !== placement.projectId) {
    throw new MigrationRefused('this generation-bearing map needs its locally configured project_id; ' +
      'do not learn that identity from the supplied map')
  }
  if (fingerprintOf(placement) === undefined) {
    throw new MigrationRefused('generation-bearing sessions need an immutable loaded placement map')
  }
  return projectId
}

/**
 * Check generations and columns; return how tables differ from the declared physical design.
 *
 * The physical design is reported rather than refused: a running application must not stop because
 * a table's sort key or index differs from the map (requirement 3.6). Columns, types and
 * generations still refuse.
 */
export async function validateGenerations(model: LogicalModel, placement: PlacementMap,
  engines: Readonly<Record<string, Engine>>, projectId: string | undefined): Promise<readonly PhysicalFinding[]> {
  if (placement.contract < GENERATIONS_SINCE) return []
  const findings: PhysicalFinding[] = []
  const project = checkMapProject(placement, projectId) as string
  if (model.version !== placement.modelVersion) {
    throw new MigrationRefused('the generation-bearing map names another session model')
  }
  for (const name of Object.keys(placement.groups).sort()) {
    const spot = placement.groups[name]!
    for (const material of [spot.source, ...spot.derived]) {
      const engine = engines[material.engine] as Engine & Partial<Fencable>
      if (typeof engine?.writeFence !== 'function' || typeof engine.validateSchema !== 'function') {
        throw new MigrationRefused(`engine ${material.engine} does not implement write generations`)
      }
      const keys: Record<string, readonly string[]> = {}
      for (const entity of Object.keys(material.layout.tables)) {
        const spec = model.entities.find((candidate) => candidate.name === entity)
        if (spec !== undefined) keys[entity] = spec.key
      }
      findings.push(...((await engine.validateSchema(material.layout, { keys })) ?? []))
      for (const table of Object.values(material.layout.tables).sort()) {
        const state = await engine.writeFence(table, { projectId: project }).state()
        if (!state.complete || state.epoch !== spot.writeEpoch) {
          throw new MigrationRefused(`the write generation for ${name} is not active in ${material.engine}; ` +
            'provision the signed map or load the current map before opening a session')
        }
      }
    }
  }
  return findings
}

export function stampValues(placement: PlacementMap, group: string, values: Readonly<Row>): Readonly<Row> {
  const epoch = placement.groups[group]?.writeEpoch
  if (epoch === undefined) return values
  if (EPOCH_COLUMN in values) throw new MigrationRefused('the write-epoch column is reserved for the SDK')
  return { ...values, [EPOCH_COLUMN]: epoch }
}

export function logicalRow(placement: PlacementMap, row: Row | null): Row | null {
  if (placement.contract >= 4 && row !== null) {
    return Object.fromEntries(Object.entries(row).filter(([key]) => key !== EPOCH_COLUMN))
  }
  return row
}

File: python/src/sde/generation.py
"""Wire vocabulary for generation-bearing maps, independent of drivers and placement classes."""

from __future__ import annotations

from collections.abc import Mapping
from typing import TYPE_CHECKING, Any, Protocol, cast

from .errors import MigrationRefused

if TYPE_CHECKING:
    from .model import LogicalModel
    from .placement import PhysicalLayout, PlacementMap
    from .session import Engine
    from .write_fence import WriteFence


class Fencable(Protocol):
    def write_fence(self, table: str, *, project_id: str) -> WriteFence: ...
    def validate_schema(self, layout: PhysicalLayout) -> None: ...


DRAIN_TABLE = "__sde_fence_drains"
EPOCH_COLUMN = "__sde_write_epoch"
MAX_EPOCH = 9_007_199_254_740_991


def check_epoch(epoch: int) -> int:
    if type(epoch) not in (int, float) or not 1 <= epoch <= MAX_EPOCH or int(epoch) != epoch:
        raise MigrationRefused("write epoch must be a positive safe integer")
    return int(epoch)


def json_numbers(value: Any) -> Any:
    """JSON has one number type. Match integral JS numbers before checking/signing a v4 map.

    Canonical encoding itself remains strict. Nonintegral/nonfinite numbers survive this pass and
    are refused by the map's canonical boundary. Earlier map contracts retain their old behavior.
    """
    if isinstance(value, float) and value.is_integer():
        return int(value)
    if isinstance(value, dict):
        return {key: json_numbers(item) for key, item in value.items()}
    if isinstance(value, list):
        return [json_numbers(item) for item in value]
    if isinstance(value, tuple):
        return tuple(json_numbers(item) for item in value)
    return value


def check_map_project(placement: PlacementMap, project_id: str | None) -> str | None:
    if placement.contract < 4:
        return None
    if project_id is None or project_id != placement.project_id:
        raise MigrationRefused(
            "this generation-bearing map needs its locally configured project_id; "
            "do not learn that identity from the supplied map"
        )
    if placement.fingerprint is None:
        raise MigrationRefused("generation-bearing sessions need an immutable loaded placement map")
    return project_id


def validate_generations(
    model: LogicalModel,
    placement: PlacementMap,
    engines: Mapping[str, Engine],
    project_id: str | None,
) -> None:
    if placement.contract < 4:
        return
    local_project = check_map_project(placement, project_id)
    assert local_project is not None
    if model.version != placement.model_version:
        raise MigrationRefused("the generation-bearing map names another session model")
    for name in sorted(placement.groups):
        spot = placement.groups[name]
        for material in spot.all():
            factory = getattr(engines[material.engine], "write_fence", None)
            if not callable(factory) or not callable(
                getattr(engines[material.engine], "validate_schema", None)
            ):
                raise MigrationRefused(
                    f"engine {material.engine} does not implement write generations"
                )
            cast(Fencable, engines[material.engine]).validate_schema(material.layout)
            for table in sorted(material.layout.tables.values()):
                state = (
                    cast(Fencable, engines[material.engine])
                    .write_fence(table, project_id=local_project)
                    .state()
                )
                if not state.complete or state.epoch != spot.write_epoch:
                    raise MigrationRefused(
                        f"the write generation for {name} is not active in {material.engine}; "
                        "provision the signed map or load the current map before opening a session"
                    )


def stamp_values(
    placement: PlacementMap, group: str, values: Mapping[str, Any]
) -> Mapping[str, Any]:
    epoch = placement.placement_of(group).write_epoch
    if epoch is None:
        return values
    if EPOCH_COLUMN in values:
        raise MigrationRefused("the write-epoch column is reserved for the SDK")
    return {**values, EPOCH_COLUMN: epoch}


def logical_row(placement: PlacementMap, row: Any) -> Any:
    if placement.contract >= 4 and isinstance(row, dict):
        return {key: value for key, value in row.items() if key != EPOCH_COLUMN}
    return row
